require decorator removes type annotations for mypy

```
$ cat ic.py  
from icontract import require

@require(lambda x: x > 0)
def f1(x: int): return x

def f2(x: int): return 0

f1("this is wrong")
f2("this is too")

$ mypy ic.py              
ic.py:9: error: Argument 1 to "f2" has incompatible type "str"; expected "int"
```

Both function calls are invalid types but mypy only sees the call as f2 as wrong.  The decorator loses the type annotations?

Related https://github.com/python/mypy/issues/3157

Hi @jamescasbon 

My bad, I indeed misread https://github.com/python/mypy/issues/3157. Mypy indeed supports  decorators which do not change the signature of the function.

Would you mind reviewing the following snippet for me? I assume that I simply need to change the argument type and return type of `icontract.require` and `icontract.ensure` argument to fix the issue?

experiment.py:
```python
from typing import TypeVar, Any, Callable

CallableT = TypeVar('CallableT', bound='Callable')

class some_decorator:
    def __init__(self, some_param: int)->None:
        return

    def __call__(self, func: CallableT) -> CallableT:
        def result(*args, **kwargs):
            print("hello")
            return func(*args, **kwargs)

        return result  # type: ignore

@some_decorator(some_param="oi")
def f(x: int, y: str)->str:
    return str(x + int(y))

f(x="oi", y=1)
```

I get the expected errors:

```bash
$ mypy experiment.py 
experiment.py:16: error: Argument "some_param" to "some_decorator" has incompatible type "str"; expected "int"
experiment.py:20: error: Argument "x" to "f" has incompatible type "str"; expected "int"
experiment.py:20: error: Argument "y" to "f" has incompatible type "int"; expected "str"
```

Ah, i redid the check with mypy --strict 

The require decorator now gives this error: 

    ic.py:3: warning: Returning Any from function declared to return "bool"

I do think this PR is a lot better - but would be very nice if we could run with strict

Hi James,
Let me have a look at it tonight (CET).

Cheers Marko

Le mer. 23 janv. 2019 à 11:05, James Casbon <notifications@github.com> a
écrit :

> Ah, i redid the check with mypy --strict
>
> The require decorator now gives this error:
>
> ic.py:3: warning: Returning Any from function declared to return "bool"
>
> I do think this PR is a lot better - but would be very nice if we could
> run with strict
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Parquery/icontract/issues/93#issuecomment-456742784>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE1ng1llR7U6d5Te2zbgDidDAmWZh7fqks5vGDP4gaJpZM4aMwNq>
> .
>

--- a/icontract/_checkers.py
+++ b/icontract/_checkers.py
@@ -229,7 +229,7 @@ class _Old:
     def __init__(self, mapping: Mapping[str, Any]) -> None:
         self.__dict__.update(mapping)
 
-    def __getattr__(self, item):
+    def __getattr__(self, item: str) -> Any:
         raise AttributeError("The snapshot with the name {!r} is not available in the OLD of a postcondition. "
                              "Have you decorated the function with a corresponding snapshot decorator?".format(item))
 

--- a/icontract/_checkers.py
+++ b/icontract/_checkers.py
@@ -4,6 +4,7 @@ import inspect
 from typing import Callable, Any, Iterable, Optional, Tuple, List, Mapping, MutableMapping, Dict
 
 import icontract._represent
+from icontract._globals import CallableT
 from icontract._types import Contract, Snapshot
 from icontract.errors import ViolationError
 
@@ -13,7 +14,7 @@ from icontract.errors import ViolationError
 # pylint: disable=unsubscriptable-object
 
 
-def _walk_decorator_stack(func: Callable[..., Any]) -> Iterable['Callable[..., Any]']:
+def _walk_decorator_stack(func: CallableT) -> Iterable['CallableT']:
     """
     Iterate through the stack of decorated functions until the original function.
 
@@ -27,9 +28,9 @@ def _walk_decorator_stack(func: Callable[..., Any]) -> Iterable['Callable[..., A
     yield func
 
 
-def find_checker(func: Callable[..., Any]) -> Optional[Callable[..., Any]]:
+def find_checker(func: CallableT) -> Optional[CallableT]:
     """Iterate through the decorator stack till we find the contract checker."""
-    contract_checker = None  # type: Optional[Callable[..., Any]]
+    contract_checker = None  # type: Optional[CallableT]
     for a_wrapper in _walk_decorator_stack(func):
         if hasattr(a_wrapper, "__preconditions__") or hasattr(a_wrapper, "__postconditions__"):
             contract_checker = a_wrapper
@@ -236,7 +237,7 @@ class _Old:
         return "a bunch of OLD values"
 
 
-def decorate_with_checker(func: Callable[..., Any]) -> Callable[..., Any]:
+def decorate_with_checker(func: CallableT) -> CallableT:
     """Decorate the function with a checker that verifies the preconditions and postconditions."""
     assert not hasattr(func, "__preconditions__"), \
         "Expected func to have no list of preconditions (there should be only a single contract checker per function)."
@@ -311,7 +312,7 @@ def decorate_with_checker(func: Callable[..., Any]) -> Callable[..., Any]:
             for contract in postconditions:
                 _assert_postcondition(contract=contract, resolved_kwargs=resolved_kwargs)
 
-        return result
+        return result  # type: ignore
 
     # Copy __doc__ and other properties so that doctests can run
     functools.update_wrapper(wrapper=wrapper, wrapped=func)
@@ -330,7 +331,7 @@ def decorate_with_checker(func: Callable[..., Any]) -> Callable[..., Any]:
     setattr(wrapper, "__postcondition_snapshots__", [])
     setattr(wrapper, "__postconditions__", [])
 
-    return wrapper
+    return wrapper  # type: ignore
 
 
 def _find_self(param_names: List[str], args: Tuple[Any, ...], kwargs: Dict[str, Any]) -> Any:
@@ -344,7 +345,7 @@ def _find_self(param_names: List[str], args: Tuple[Any, ...], kwargs: Dict[str, 
     return instance
 
 
-def _decorate_with_invariants(func: Callable[..., Any], is_init: bool) -> Callable[..., Any]:
+def _decorate_with_invariants(func: CallableT, is_init: bool) -> CallableT:
     """
     Decorate the function ``func`` of the class ``cls`` with invariant checks.
 
@@ -391,7 +392,7 @@ def _decorate_with_invariants(func: Callable[..., Any], is_init: bool) -> Callab
 
     setattr(wrapper, "__is_invariant_check__", True)
 
-    return wrapper
+    return wrapper  # type: ignore
 
 
 class _DummyClass:
@@ -403,7 +404,7 @@ class _DummyClass:
 _SLOT_WRAPPER_TYPE = type(_DummyClass.__init__)  # pylint: disable=invalid-name
 
 
-def _already_decorated_with_invariants(func: Callable[..., Any]) -> bool:
+def _already_decorated_with_invariants(func: CallableT) -> bool:
     """Check if the function has been already decorated with an invariant check by going through its decorator stack."""
     already_decorated = False
     for a_decorator in _walk_decorator_stack(func=func):
@@ -467,9 +468,9 @@ def add_invariant_checks(cls: type) -> None:
         setattr(cls, name, wrapper)
 
     for name, prop in names_properties:
-        fget = _decorate_with_invariants(func=prop.fget, is_init=False) if prop.fget else None
-        fset = _decorate_with_invariants(func=prop.fset, is_init=False) if prop.fset else None
-        fdel = _decorate_with_invariants(func=prop.fdel, is_init=False) if prop.fdel else None
-
-        new_prop = property(fget=fget, fset=fset, fdel=fdel, doc=prop.__doc__)
+        new_prop = property(  # type: ignore
+            fget=_decorate_with_invariants(func=prop.fget, is_init=False) if prop.fget else None,
+            fset=_decorate_with_invariants(func=prop.fset, is_init=False) if prop.fset else None,
+            fdel=_decorate_with_invariants(func=prop.fdel, is_init=False) if prop.fdel else None,
+            doc=prop.__doc__)
         setattr(cls, name, new_prop)

--- a/icontract/_decorators.py
+++ b/icontract/_decorators.py
@@ -23,7 +23,7 @@ class require:  # pylint: disable=invalid-name
                  description: Optional[str] = None,
                  a_repr: reprlib.Repr = icontract._globals.aRepr,
                  enabled: bool = __debug__,
-                 error: Union[Callable[..., Exception], type] = None) -> None:
+                 error: Optional[Union[Callable[..., Exception], type]] = None) -> None:
         """
         Initialize.
 
@@ -184,7 +184,7 @@ class ensure:  # pylint: disable=invalid-name
                  description: Optional[str] = None,
                  a_repr: reprlib.Repr = icontract._globals.aRepr,
                  enabled: bool = __debug__,
-                 error: Union[Callable[..., Exception], type] = None) -> None:
+                 error: Optional[Union[Callable[..., Exception], type]] = None) -> None:
         """
         Initialize.
 
@@ -266,7 +266,7 @@ class invariant:  # pylint: disable=invalid-name
                  description: Optional[str] = None,
                  a_repr: reprlib.Repr = icontract._globals.aRepr,
                  enabled: bool = __debug__,
-                 error: Union[Callable[..., Exception], type] = None) -> None:
+                 error: Optional[Union[Callable[..., Exception], type]] = None) -> None:
         """
         Initialize a class decorator to establish the invariant on all the public methods.
 

--- a/icontract/_decorators.py
+++ b/icontract/_decorators.py
@@ -2,6 +2,7 @@
 import reprlib
 from typing import Callable, Optional, Union, Any, List  # pylint: disable=unused-import
 
+from icontract._globals import CallableT
 from icontract._types import Contract, Snapshot
 
 import icontract._checkers
@@ -54,7 +55,7 @@ class require:  # pylint: disable=invalid-name
 
         self._contract = Contract(condition=condition, description=description, a_repr=a_repr, error=error)
 
-    def __call__(self, func: Callable[..., Any]) -> Callable[..., Any]:
+    def __call__(self, func: CallableT) -> CallableT:
         """
         Add the precondition to the list of preconditions of the function ``func``.
 
@@ -131,7 +132,7 @@ class snapshot:  # pylint: disable=invalid-name
         if enabled:
             self._snapshot = Snapshot(capture=capture, name=name)
 
-    def __call__(self, func: Callable[..., Any]) -> Callable[..., Any]:
+    def __call__(self, func: CallableT) -> CallableT:
         """
         Add the snapshot to the list of snapshots of the function ``func``.
 
@@ -215,7 +216,7 @@ class ensure:  # pylint: disable=invalid-name
 
         self._contract = Contract(condition=condition, description=description, a_repr=a_repr, error=error)
 
-    def __call__(self, func: Callable[..., Any]) -> Callable[..., Any]:
+    def __call__(self, func: CallableT) -> CallableT:
         """
         Add the postcondition to the list of postconditions of the function ``func``.
 

--- a/icontract/_decorators.py
+++ b/icontract/_decorators.py
@@ -19,7 +19,7 @@ class require:  # pylint: disable=invalid-name
 
     # pylint: disable=too-many-instance-attributes
     def __init__(self,
-                 condition: Callable[..., bool],
+                 condition: Callable[..., Any],
                  description: Optional[str] = None,
                  a_repr: reprlib.Repr = icontract._globals.aRepr,
                  enabled: bool = __debug__,
@@ -180,7 +180,7 @@ class ensure:  # pylint: disable=invalid-name
 
     # pylint: disable=too-many-instance-attributes
     def __init__(self,
-                 condition: Callable[..., bool],
+                 condition: Callable[..., Any],
                  description: Optional[str] = None,
                  a_repr: reprlib.Repr = icontract._globals.aRepr,
                  enabled: bool = __debug__,
@@ -262,7 +262,7 @@ class invariant:  # pylint: disable=invalid-name
     """
 
     def __init__(self,
-                 condition: Callable[..., bool],
+                 condition: Callable[..., Any],
                  description: Optional[str] = None,
                  a_repr: reprlib.Repr = icontract._globals.aRepr,
                  enabled: bool = __debug__,

--- a/icontract/_globals.py
+++ b/icontract/_globals.py
@@ -2,11 +2,13 @@
 
 import os
 import reprlib
+from typing import TypeVar, Callable
 
 # Default representation instance.
 #
 # The limits are set way higher than reprlib.aRepr since the default reprlib limits are not suitable for
 # the production systems.
+
 aRepr = reprlib.Repr()  # pylint: disable=invalid-name
 aRepr.maxdict = 50
 aRepr.maxlist = 50
@@ -25,3 +27,4 @@ aRepr.maxother = 256
 #
 # Contracts marked with SLOW are also disabled if the interpreter is run in optimized mode (``-O`` or ``-OO``).
 SLOW = __debug__ and os.environ.get("ICONTRACT_SLOW", "") != ""
+CallableT = TypeVar('CallableT', bound='Callable')

--- a/icontract/_types.py
+++ b/icontract/_types.py
@@ -17,7 +17,7 @@ class Contract:
                  condition: Callable[..., bool],
                  description: Optional[str] = None,
                  a_repr: reprlib.Repr = icontract._globals.aRepr,
-                 error: Union[Callable[..., Exception], type] = None) -> None:
+                 error: Optional[Union[Callable[..., Exception], type]] = None) -> None:
         """
         Initialize.
 

--- a/tests/test_mypy_decorators.py
+++ b/tests/test_mypy_decorators.py
@@ -8,40 +8,46 @@ import unittest
 
 class TestMypyDecorators(unittest.TestCase):
     def test_mypy_me(self):
-        with tempfile.NamedTemporaryFile(prefix="mypy_fail_case", suffix=".py") as tmp:
+        with tempfile.NamedTemporaryFile(prefix="mypy_fail_case_", suffix=".py") as tmp:
             tmp.file.write(
                 textwrap.dedent('''\
                 """Implement a fail case for mypy to test that the types are preserved with the decorators."""
 
                 import icontract
-                
-                @icontract.require(lambda x: x > 0)
-                def f1(x: int):
+
+                def some_precond(x: int) -> bool: return x > 0
+
+                @icontract.require(some_precond)
+                def f1(x: int) -> int:
                     return x
                 f1("this is wrong")
-                
-                @icontract.ensure(lambda result: result > 0)
-                def f2(x: int):
+
+                def some_postcond(result: int) -> bool: return result > 0
+
+                @icontract.ensure(some_postcond)
+                def f2(x: int) -> int:
                     return x
                 f2("this is wrong")
-                
-                @icontract.snapshot(lambda x: x)
-                def f3(x: int):
+
+                def some_snapshot(x: int) -> int: return x
+
+                @icontract.snapshot(some_snapshot)
+                def f3(x: int) -> int:
                     return x
                 f3("this is wrong")
                 ''').encode())
 
             tmp.file.flush()
 
-            proc = subprocess.Popen(['mypy', tmp.name], universal_newlines=True, stdout=subprocess.PIPE)
+            proc = subprocess.Popen(['mypy', '--strict', tmp.name], universal_newlines=True, stdout=subprocess.PIPE)
             out, err = proc.communicate()
 
             self.assertIsNone(err)
             self.assertEqual(
                 textwrap.dedent('''\
-                    {path}:8: error: Argument 1 to "f1" has incompatible type "str"; expected "int"
-                    {path}:13: error: Argument 1 to "f2" has incompatible type "str"; expected "int"
-                    {path}:18: error: Argument 1 to "f3" has incompatible type "str"; expected "int"
+                    {path}:10: error: Argument 1 to "f1" has incompatible type "str"; expected "int"
+                    {path}:17: error: Argument 1 to "f2" has incompatible type "str"; expected "int"
+                    {path}:24: error: Argument 1 to "f3" has incompatible type "str"; expected "int"
                     '''.format(path=tmp.name)),
                 out)
 

--- a/tests/test_mypy_decorators.py
+++ b/tests/test_mypy_decorators.py
@@ -15,23 +15,17 @@ class TestMypyDecorators(unittest.TestCase):
 
                 import icontract
 
-                def some_precond(x: int) -> bool: return x > 0
-
-                @icontract.require(some_precond)
+                @icontract.require(lambda x: x > 0)
                 def f1(x: int) -> int:
                     return x
                 f1("this is wrong")
 
-                def some_postcond(result: int) -> bool: return result > 0
-
-                @icontract.ensure(some_postcond)
+                @icontract.ensure(lambda result: result > 0)
                 def f2(x: int) -> int:
                     return x
                 f2("this is wrong")
 
-                def some_snapshot(x: int) -> int: return x
-
-                @icontract.snapshot(some_snapshot)
+                @icontract.snapshot(lambda x: x)
                 def f3(x: int) -> int:
                     return x
                 f3("this is wrong")
@@ -45,9 +39,9 @@ class TestMypyDecorators(unittest.TestCase):
             self.assertIsNone(err)
             self.assertEqual(
                 textwrap.dedent('''\
-                    {path}:10: error: Argument 1 to "f1" has incompatible type "str"; expected "int"
-                    {path}:17: error: Argument 1 to "f2" has incompatible type "str"; expected "int"
-                    {path}:24: error: Argument 1 to "f3" has incompatible type "str"; expected "int"
+                    {path}:8: error: Argument 1 to "f1" has incompatible type "str"; expected "int"
+                    {path}:13: error: Argument 1 to "f2" has incompatible type "str"; expected "int"
+                    {path}:18: error: Argument 1 to "f3" has incompatible type "str"; expected "int"
                     '''.format(path=tmp.name)),
                 out)
 

--- a/tests/test_mypy_decorators.py
+++ b/tests/test_mypy_decorators.py
@@ -1,0 +1,50 @@
+# pylint: disable=missing-docstring
+
+import subprocess
+import tempfile
+import textwrap
+import unittest
+
+
+class TestMypyDecorators(unittest.TestCase):
+    def test_mypy_me(self):
+        with tempfile.NamedTemporaryFile(prefix="mypy_fail_case", suffix=".py") as tmp:
+            tmp.file.write(
+                textwrap.dedent('''\
+                """Implement a fail case for mypy to test that the types are preserved with the decorators."""
+
+                import icontract
+                
+                @icontract.require(lambda x: x > 0)
+                def f1(x: int):
+                    return x
+                f1("this is wrong")
+                
+                @icontract.ensure(lambda result: result > 0)
+                def f2(x: int):
+                    return x
+                f2("this is wrong")
+                
+                @icontract.snapshot(lambda x: x)
+                def f3(x: int):
+                    return x
+                f3("this is wrong")
+                ''').encode())
+
+            tmp.file.flush()
+
+            proc = subprocess.Popen(['mypy', tmp.name], universal_newlines=True, stdout=subprocess.PIPE)
+            out, err = proc.communicate()
+
+            self.assertIsNone(err)
+            self.assertEqual(
+                textwrap.dedent('''\
+                    {path}:8: error: Argument 1 to "f1" has incompatible type "str"; expected "int"
+                    {path}:13: error: Argument 1 to "f2" has incompatible type "str"; expected "int"
+                    {path}:18: error: Argument 1 to "f3" has incompatible type "str"; expected "int"
+                    '''.format(path=tmp.name)),
+                out)
+
+
+if __name__ == '__main__':
+    unittest.main()

